### PR TITLE
Expose query and querier accept_replies option

### PR DESCRIPTION
**Pull request by DenisBiryukov91**
## Description

Expose query and querier accept_replies option

### What does this PR do?

Expose query and querier accept_replies option and z_query_accepts_replies() getter

### Why is this change needed?

To align with Zenoh Rust API

### Related Issues

---
## 🏷️ Label-Based Checklist

Based on the labels applied to this PR, please complete these additional requirements:

**Labels:** `new feature`

## 🆕 New Feature Requirements

Since this PR adds a new feature:

- [x] **Feature scope documented** - Clear description of what the feature does and why it's needed
- [x] **Minimum necessary code** - Implementation is as simple as possible, doesn't overcomplicate the system
- [x] **New APIs well-designed** - Public APIs are intuitive, consistent with existing APIs
- [x] **Comprehensive tests** - All functionality is tested (happy path + edge cases + error cases)
- [x] **Examples provided** - Usage examples in code comments or separate example files
- [x] **Documentation added** - New docs explaining the feature, its use cases, and API
- [x] **Feature flag considered** - Can the feature be enabled/disabled for gradual rollout?
- [x] **Performance impact assessed** - Memory, CPU, storage implications measured
- [x] **Integration tested** - Feature works with existing features

**Consider:** Can this feature be split into smaller, incremental PRs?

**Instructions:**
1. Check off items as you complete them (change `- [ ]` to `- [x]`)
2. The PR checklist CI will verify these are completed

*This checklist updates automatically when labels change, but preserves your checked boxes.*

### CODE REVIEW

File: include/zenoh-pico/api/constants.h [hunk 1]
```diff
@@ -256,6 +256,32 @@ typedef enum {
...
+ * The kinds of accepted query replies.
+ *
+ * The queryable may serve glob-like key expressions.
+ * E.g., the queryable may be declared with the key expression `foo/ *`.
+ * At the same time, it may send replies with more specific key expressions, e.g., `foo/bar` or `foo/baz`.
+ * This may cause a situation when the queryable receives a query with the key expression `foo/bar`
+ * and replies to it with the key expression `foo/baz`.
+ * By default, this behavior is not allowed. Calling `z_query_reply` value on a query for `foo/bar` with key expression
+ * `foo/baz` will result in an error on the sending side. But if the query is sent with the `accept_replies` flag set to
+ * `Z_REPLY_KEYEXPR_ANY` in either `z_get_options_t` or `z_querier_options_t`, then the reply with a disjoint key
```
Comment (milyin) [inline include/zenoh-pico/api/constants.h:278]:
```suggestion
 * `Z_REPLY_KEYEXPR_ANY` is either `z_get_options_t` or `z_querier_options_t`, then the reply with a disjoint key
```

File: include/zenoh-pico/api/constants.h [hunk 2]
```diff
@@ -256,6 +256,32 @@ typedef enum {
...
+ * Enumerators:
+ *   Z_REPLY_KEYEXPR_ANY: Accept replies on any key expression.
+ *   Z_REPLY_KEYEXPR_MATCHING_QUERY: Accept replies only to intersecting key expressions intersecting with query's own
+ * key expression.
+ */
+typedef enum z_reply_keyexpr_t {
+    Z_REPLY_KEYEXPR_ANY = 0,
+    Z_REPLY_KEYEXPR_MATCHING_QUERY = 1,
+} z_reply_keyexpr_t;
+#define Z_REPLY_KEYEXPR_DEFAULT Z_REPLY_KEYEXPR_MATCHING_QUERY
```
Comment (milyin) [inline include/zenoh-pico/api/constants.h]: zenoh-c doesn't provide this constant. Should we add it there also? Or is function `z_reply_keyexpr_default` enough?

**Review by milyin — APPROVED** [review]

File: tests/z_api_queryable_test.c [hunk 3]
```diff
@@ -77,8 +72,377 @@ void test_queryable_keyexpr(void) {
...
+        z_owned_config_t config2;
+        z_config_default(&config2);
+        zp_config_insert(z_loan_mut(config2), Z_CONFIG_MODE_KEY, "client");
+        if (z_open(s2, z_move(config2), NULL) < 0) {
+            printf("Unable to open second session for local queryable test!\n");
+            z_drop(z_move(*s));
+            exit(-1);
+        }
+    }
+}
```
[the remark this answers is not in the document]
Comment (DenisBiryukov91) [reply tests/z_api_queryable_test.c:97]: Read and lease tasks are started automatically by default

File: include/zenoh-pico/net/primitives.h [hunk 4]
```diff
@@ -324,7 +330,8 @@ z_result_t _z_query(const _z_session_rc_t *session, _z_optional_id_t querier_id,
                     z_consolidation_mode_t consolidation, _z_bytes_t *payload, _z_encoding_t *encoding,
                     _z_closure_reply_callback_t callback, _z_drop_handler_t dropper, void *arg, uint64_t timeout_ms,
                     _z_bytes_t *attachment, _z_n_qos_t qos, _z_source_info_t *source_info,
-                    z_locality_t allowed_destination, _z_cancellation_token_rc_t *opt_cancellation_token);
+                    z_reply_keyexpr_t accept_replies, z_locality_t allowed_destination,
+                    _z_cancellation_token_rc_t *opt_cancellation_token);
```
[the remark this answers is not in the document]
Comment (DenisBiryukov91) [reply include/zenoh-pico/net/primitives.h:334]: fixed

File: src/protocol/codec/message.c [hunk 5]
```diff
@@ -479,6 +494,7 @@ z_result_t _z_query_decode(_z_msg_query_t *msg, _z_zbuf_t *zbf, uint8_t header)
     }
     if (_Z_HAS_FLAG(header, _Z_FLAG_Z_Q_P)) {
         _Z_RETURN_IF_ERR(_z_slice_decode(&msg->_parameters, zbf));
+        msg->_anyke = _z_parameters_has_anyke((const char *)msg->_parameters.start, msg->_parameters.len);
     } else {
         _z_slice_clear(&msg->_parameters);
```
[the remark this answers is not in the document]
Comment (DenisBiryukov91) [reply src/protocol/codec/message.c:499]: fixed

File: include/zenoh-pico/api/types.h [hunk 6]
```diff
@@ -490,6 +493,7 @@ typedef struct {
     z_source_info_t *source_info;
     z_moved_cancellation_token_t *cancellation_token;
 #endif
+    z_reply_keyexpr_t accept_replies;
 } z_get_options_t;
```
[the remark this answers is not in the document]
Comment (DenisBiryukov91) [reply include/zenoh-pico/api/types.h:497]: fixed

File: include/zenoh-pico/api/constants.h [hunk 7]
```diff
@@ -256,6 +256,32 @@ typedef enum {
...
+ * `foo/baz` will result in an error on the sending side. But if the query is sent with the `accept_replies` flag set to
+ * `Z_REPLY_KEYEXPR_ANY` in either `z_get_options_t` or `z_querier_options_t`, then the reply with a disjoint key
+ * expression will be accepted for this query.
+ *
+ * The queryable may check wether disjoint replies are allowed for a query with `z_query_accepts_replies` function.
+ *
+ * Enumerators:
+ *   Z_REPLY_KEYEXPR_ANY: Accept replies on any key expression.
+ *   Z_REPLY_KEYEXPR_MATCHING_QUERY: Accept replies only to intersecting key expressions intersecting with query's own
+ * key expression.
```
[the remark this answers is not in the document]
Comment (DenisBiryukov91) [reply include/zenoh-pico/api/constants.h:286]: fixed

File: tests/z_msgcodec_test.c [hunk 8]
```diff
@@ -1391,6 +1405,39 @@ void query_message(void) {
...
+            z_result_t res = _z_query_decode(&decoded, &zbf, header);
+            assert(_Z_RES_OK == res);
+            if (anyke) {
+                // SAFETY: only use null-terminated parameters in tests.
+                // Flawfinder: ignore [CWE-126]
+                size_t params_len = params[i] == NULL ? 0 : strlen(params[i]);
+                assert(decoded._anyke);
+                assert(_z_parameters_has_anyke((const char *)decoded._parameters.start, decoded._parameters.len));
+                assert(params_len <= decoded._parameters.len);
+                assert(strncmp(params[i], (const char *)decoded._parameters.start, params_len) == 0);
```
[the remark this answers is not in the document]
Comment (DenisBiryukov91) [reply tests/z_msgcodec_test.c]: fixed

File: src/protocol/codec/message.c [hunk 9]
```diff
@@ -394,8 +395,8 @@ z_result_t _z_query_encode(_z_wbuf_t *wbf, const _z_msg_query_t *msg) {
     z_result_t ret = _Z_RES_OK;
     uint8_t header = _Z_MID_Z_QUERY;
 
-    bool has_params = _z_slice_check(&msg->_parameters);
-    if (has_params) {
+    bool has_params = _z_slice_check(&msg->_parameters) && msg->_parameters.len > 0;
```
Comment (steils) [inline src/protocol/codec/message.c:398]: It will change the serialization when an explicit empty parameter string is passed.
I.e. `z_get(..., "", ...)` will be serialized as "no parameters" (not an empty slice for the existing parameter field). But in the local path it is still a present but empty parameter field (see  [src/session/loopback.c:131]).

File: src/protocol/codec/message.c [hunk 10]
```diff
@@ -410,9 +411,22 @@ z_result_t _z_query_encode(_z_wbuf_t *wbf, const _z_msg_query_t *msg) {
     if (has_consolidation) {
         _Z_RETURN_IF_ERR(_z_uint8_encode(wbf, msg->_consolidation));
     }
-    if (has_params) {
+    if (msg->_anyke) {
```
Comment (steils) [inline src/protocol/codec/message.c]: Here when `accept_replies = Z_REPLY_KEYEXPR_ANY`, "_anyke" is appended to the parameter sting. But this code isn't called for local queries. Therefore the same query can have parameters "foo=bar;_anyke" remotely but "foo=bar" locally.

**Review by steils — CHANGES_REQUESTED** [review]

File: tests/z_api_queryable_test.c [hunk 11]
```diff
@@ -77,8 +72,377 @@ void test_queryable_keyexpr(void) {
...
+        // 2b: reply on a key expression NOT matching the querier key — should be rejected
+        z_view_keyexpr_t querier_ke2;
+        z_view_keyexpr_from_str(&querier_ke2, "test/querier_accept/d");
+
+        z_querier_options_t querier_opts2;
+        z_querier_options_default(&querier_opts2);
+        querier_opts2.accept_replies = Z_REPLY_KEYEXPR_MATCHING_QUERY;
+
+        z_owned_querier_t querier2;
+        z_declare_querier(z_loan(s), &querier2, z_loan(querier_ke2), &querier_opts2);
```
Comment (steils) [inline tests/z_api_queryable_test.c]: not `s2`?

File: src/protocol/codec/message.c (hunk 10 above)
Comment (DenisBiryukov91) [reply src/protocol/codec/message.c]: Yes, this is ok I believe, since parameters starting from `_` are reserved parameters, so user should ignore them/do not have any expectations regarding their presence/absence. As of today in zenoh-pico, we never explicitly look for _anyke parameter in the parameters list but use _anyke field of the query which in these both cases is correctly set. 
The data sent outside will always have _anyke in the parameters list.

File: src/protocol/codec/message.c (hunk 9 above)
Comment (DenisBiryukov91) [reply src/protocol/codec/message.c:398]: yes, but upon deserialization zero-length params or no-params cases will produce the same empty _z_slice_t, so undistinguishable

File: tests/z_api_queryable_test.c (hunk 11 above)
Comment (DenisBiryukov91) [reply tests/z_api_queryable_test.c]: Fixed

**Review by steils — APPROVED** [review]
